### PR TITLE
Added Commands, Fixed Trunk/Frunk

### PR DESCRIPTION
**Pull request by boxystudio**
**New Commands:**

- Window Control
- Software Update (if available)
- Frunk Actuation

**Fixed Commands:**

- Trunk Actuation

### CODE REVIEW

File: src/StephanGroen/Tesla/Tesla.php [hunk 1]
```diff
@@ -89,6 +89,11 @@ public function wakeUp() : array
         return $this->sendRequest('/wake_up', [], 'POST')['response'];
     }
 
+    public function startSoftwareUpdate() : array
+    {
+        return $this->sendRequest('/command/schedule_software_update', [ 'offset_sec' => 0 ], 'POST')['response'];
+    }
+
```
Comment (corsair) [inline src/StephanGroen/Tesla/Tesla.php]: The offset seconds should be able to be set as a variable and default to 0 (immediate), to allow for delayed installations.

File: src/StephanGroen/Tesla/Tesla.php [hunk 2]
```diff
@@ -157,6 +162,11 @@ public function lockDoors() : array
         return $this->sendRequest('/command/door_lock', [], 'POST')['response'];
     }
 
+    public function windowControl(string $state = 'close') : array
+    {
+        return $this->sendRequest('/command/window_control', [ 'command' => $state, 'lat' => 0, 'lon' => 0 ], 'POST')['response'];
+    }
+
```
Comment (corsair) [inline src/StephanGroen/Tesla/Tesla.php]: We want to ensure that the window control command can receive latitude and longitude variables (defaulting to 0), as this is required for "close" but ignored for "vent".

**Comment by corsair** [comment]
Thank you for your contribution, @boxystudio! I left some comments regarding some changes that would likely be beneficial to ensure full usage of the commands added.

Let me know your thoughts!

File: src/StephanGroen/Tesla/Tesla.php (hunk 1 above)
Comment (boxystudio) [reply src/StephanGroen/Tesla/Tesla.php]: Good idea! Added a new commit with this change.

**Comment by corsair** [comment]
@boxystudio just checking in! Awaiting the windowControl function change to allow for this to behave properly, otherwise the usage of "close" would not work without the vehicle's current latitude/longitude.

More context:
https://github.com/timdorr/tesla-api/commit/e8ea847d73592c7aedcea83a11a41e9486600141

**Review by corsair — CHANGES_REQUESTED** [review]

**Comment by boxystudio** [comment]
Pushed an update to resolve this issue.

**Review by corsair — COMMENTED** [review]
Looks great! Thank you for making this change.

**Comment by Kozzirago** [comment]
Does anybody know if that API still works on the new Model 3 Highland?

**Pull request merged**